File: leetcode/july-challenge/week-3/CourseScheduling2.cxx
class Solution {
public:
    vector<int> findOrder(int numCourses, vector<vector<int>>& prerequisites) {
        vector<vector<int>> g(numCourses);
        vector<int> inorder(numCourses, 0);
        vector<int> res;
        queue<int> q;
        int cnt = 0;
        
        for (int i = 0; i < prerequisites.size(); ++i) {
            int a = prerequisites[i][0];
            int b = prerequisites[i][1];
            
            g[b].push_back(a);
        }
        
        for (int i = 0; i < numCourses; ++i) {
            for (int j = 0; j < g[i].size(); ++j) {
                ++inorder[g[i][j]];
            }
        }
        
        for (int i = 0; i < numCourses; ++i) {
            if (inorder[i] == 0) {
                q.push(i);
            }
        }
        
        while (q.size() > 0) {
            int node = q.front();
            res.push_back(node);
            q.pop();
            for (int j = 0; j < g[node].size(); ++j) {
                --inorder[g[node][j]];
                if (inorder[g[node][j]] == 0) {
                    q.push(g[node][j]);
                }
            }
            ++cnt;
        }
        
        
        return cnt == numCourses ? res : vector<int>();
    }
};

File: leetcode/july-challenge/week-3/CourseScheduling2.java
class Solution {
    public int[] findOrder(int numCourses, int[][] prerequisites) {
        Map<Integer, List<Integer>> graph = new HashMap<>();
        Queue<Integer> queue = new LinkedList<>();
        int[] inorder = new int[numCourses];
        int[] topoSort = new int[numCourses];
        
        for (int[] e : prerequisites) {
            int src = e[1];
            int dest = e[0];
            List<Integer> list = graph.getOrDefault(src, new ArrayList<Integer>());
            list.add(dest);
            graph.put(src, list);
            
            ++inorder[dest];
        }
        
        for (int i = 0; i < numCourses; ++i) {
            if (inorder[i] == 0) {
                queue.add(i);
            }
        }
        
        int i = 0;
        while(!queue.isEmpty()) {
            int node = queue.poll();
            topoSort[i++] = node;
            if (graph.containsKey(node)) {
                for (int j : graph.get(node)) {
                    --inorder[j];
                    if (inorder[j] == 0) {
                        queue.add(j);
                    }
                }
            }
        }
        
        return (i == numCourses) ? topoSort : new int[0];
        
    }
}
